### PR TITLE
Energy system

### PR DESCRIPTION
**Pull request by Paper-2**
# Description

Adds a new system that adds a cost to cards.

## Related issue(s)

https://github.com/Saplings-Projects/1M_sub/issues/56

## List of changes

A more detailed list of the changes.
- The current amount of energy is displayed on the testing scene
- Cards have a cost tied to them
- A card can only be selected if the current energy meets its cost
- A new effect to add more energy
- A var to allow cheating for testing purposes

## Tests

Energy generation

Having no energy

Having energy

Adding energy

Using energy

### CODE REVIEW

**Comment by Paper-2** [comment]
Test are failing. Reason being Player does not exist in PlayerManager. It could be solved by either not using PlayerManager to use the component or adding a player to PlayerManager in the tests.

File: Entity/Components/EnergyComponent.gd [hunk 1]
```diff
@@ -0,0 +1,38 @@
+extends EntityComponent
+class_name EnergyComponent
+
+@export var max_energy: int = 4
```
Comment (Turtyo) [inline Entity/Components/EnergyComponent.gd:4]: capitalize this as it won't really change usually

File: Entity/Components/EnergyComponent.gd [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+	if card != null:
+		return energy >= card.card_data.energy_cost
+
+	return true
+
+
+func use_energy(card: CardBase) -> void:
+	energy -= card.energy_cost
+
+	if energy < 0:
```
Comment (Turtyo) [inline Entity/Components/EnergyComponent.gd:25]: this might hide a problem (energy shouldn't be negative)
instead of setting the energy back to 0, we should assert the energy is not lower than 0 instead

File: Entity/Components/EnergyComponent.gd [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+	on_energy_changed.emit(energy)
+
+
+func on_turn_end() -> void:
+	if energy >= max_energy:
+		energy = max_energy
+		
+	else:
+		energy += energy_generation
```
Comment (Turtyo) [inline Entity/Components/EnergyComponent.gd:36]: this should be the other way
add the energy generation and then set it back to the max if it's too big
because right now, if i have 3 energy i will start my next turn with 6 (since 3 is not the max of 4)

File: Tests/TestBase.gd [hunk 4]
```diff
@@ -45,7 +48,8 @@ func before_each():
 	_player_stat_component.get_stats().ready_entity_stats()
 	_enemy_stat_component.get_stats().ready_entity_stats()
 	_enemy_2_stat_component.get_stats().ready_entity_stats()
-	
+
+	_player_energy_component.ignore_cost = true
```
Comment (Turtyo) [inline Tests/TestBase.gd:52]: why ? this is a test, you shouldn't ignore card cost

File: Entity/Player/Player.gd [hunk 5]
```diff
@@ -6,7 +6,12 @@ class_name Player
...
 	PlayerManager.set_player(self)
+	get_energy_component().init_entity_component(self)
 
 
 func _exit_tree() -> void:
 	PlayerManager.set_player(null)
+
+#this PR is going to be so easy
+func get_energy_component() -> EnergyComponent:
+	return Helpers.get_first_child_node_of_type(self, EnergyComponent) as EnergyComponent
```
Comment (Turtyo) [inline Entity/Player/Player.gd:17]: add a line at the end of your file please

File: UI/EnergyLabel.gd [hunk 6]
```diff
@@ -0,0 +1,30 @@
+extends Label
```
Comment (Turtyo) [inline UI/EnergyLabel.gd:1]: i'm not familiar with label, what does it do ?

File: Entity/Player/Player.gd [hunk 7]
```diff
@@ -6,7 +6,12 @@ class_name Player
 func _ready() -> void:
 	super()
 	PlayerManager.set_player(self)
+	get_energy_component().init_entity_component(self)
```
Comment (Turtyo) [inline Entity/Player/Player.gd:9]: is this needed ? I know that for health we don't need to do it since it's not in the code anywhere, why do you need to do it for the energy component ?

**Review by Turtyo — CHANGES_REQUESTED** [review]

File: Tests/TestBase.gd (hunk 4 above)
Comment (Paper-2) [reply Tests/TestBase.gd:52]: Now that I think about it, its not doing anything right now. Players first need to queue the card and then play the card.
I'm guessing the cards that are used in the tests skip queuing the cards.

File: Entity/Player/Player.gd (hunk 7 above)
Comment (Turtyo) [reply Entity/Player/Player.gd]: if it's redundant it should be removed

File: Tests/TestBase.gd (hunk 4 above)
Comment (Turtyo) [reply Tests/TestBase.gd]: you don't interact physically with the cards indeed, so they are not queued

Comment (Turtyo) [reply Tests/TestBase.gd]: ah the idea was the allow other tests to continue working without having to worry about energy cost ?

Comment (Paper-2) [reply Tests/TestBase.gd]: Yes, those tests are testing for other stuff and not the energy system directly.

Comment (Turtyo) [reply Tests/TestBase.gd]: yes so it's logical to add this line indeed, just need to overwrite that in the init of the energy tests

File: Cards/EnergyData.gd [hunk 8]
```diff
@@ -0,0 +1,4 @@
+extends Resource
+class_name EnergyData
+
+@export var energy_cost: int = 0
```
Comment (Turtyo) [inline Cards/EnergyData.gd:4]: add a line at the end of the file please

File: Entity/Components/EnergyComponent.gd [hunk 9]
```diff
@@ -1,7 +1,7 @@
 extends EntityComponent
 class_name EnergyComponent
 
-@export var max_energy: int = 4
+@export var Max_energy: int = 4
```
Comment (Turtyo) [inline Entity/Components/EnergyComponent.gd:4]: by capitalize I meant everything
`MAX_ENERGY`

File: Entity/Components/EnergyComponent.gd [hunk 10]
```diff
@@ -23,20 +23,20 @@ func use_energy(card: CardBase) -> void:
 	energy -= card.energy_cost
 
 	if energy < 0:
-		energy = 0
+		push_error("energy cannot be negative")
 
 	on_energy_changed.emit()
 
 
 func on_turn_end() -> void:
```
Comment (Turtyo) [inline Entity/Components/EnergyComponent.gd:32]: this should probably be a `on_turn_start`
reason for that all the other stuff happens at the start of turns. Also, we might have an effect that reduces energy at the start of turns, so it's more logical to put the logic of energy generation at the start
Another problem with that is that since earlier you wrote `starting_energy = 0` and the energy is generated at the end of turns, it means the player has 0 energy for the first turn
Edit : I see start energy was changed to 3, but this doesn't change the idea of putting everything at the start of the turn and start with 0 energy

TL;DR : make this a `on_turn_start`

File: Entity/Components/EnergyComponent.gd [hunk 11]
```diff
@@ -23,20 +23,20 @@ func use_energy(card: CardBase) -> void:
...
+	if energy >= Max_energy:
+		energy = Max_energy
+
 	on_energy_changed.emit()
 
 func add_energy(value: int) -> void:
 	energy += value
-	print(energy)
\ No newline at end of file
+	print(energy)
```
Comment (Turtyo) [inline Entity/Components/EnergyComponent.gd:43]: is there a reason for printing the energy ?

File: Tests/test_energy.gd [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+func test_add_energy():
+	card = load("res://Cards/Resource/Card_Energy.tres") # adds 70 energy
+
+	card.on_card_play(_player, [_player])
+	assert_eq(_player_energy_component.energy, 70)
+
+
+func test_using_energy():
+	_player_energy_component.add_energy(10)
+	card = load("res://Cards/Resource/Card_Damage.tres") # costs 7 by default
```
Comment (Turtyo) [inline Tests/test_energy.gd:22]: does it cost 7 by default or do you want to change it to 7 by default in this test ?
if it's 7 by default, it's too high and it needs to be realistic, you change it to 7 only in this test.

File: Tests/test_energy.gd [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+
+	card.on_card_play(PlayerManager.player, [_enemy])
+
+	assert_eq(_player_energy_component.energy, 3)
+
+func test_no_energy():
+
+	_card_container.draw_cards(1)
+
+	var _queued_card: CardWorld = _card_container.cards_in_hand[0]
```
Comment (Turtyo) [inline Tests/test_energy.gd:33]: name it `card_to_queue` because this name makes it feel like it's already being queued

File: Tests/test_energy.gd [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+func test_using_energy():
+	_player_energy_component.add_energy(10)
+	card = load("res://Cards/Resource/Card_Damage.tres") # costs 7 by default
+	card.energy.energy_cost = 7
+
+	card.on_card_play(PlayerManager.player, [_enemy])
+
+	assert_eq(_player_energy_component.energy, 3)
+
+func test_no_energy():
```
Comment (Turtyo) [inline Tests/test_energy.gd:29]: `test_not_enough_energy` would probably be a lot better
missing void return type for all test functions btw

File: Tests/test_energy.gd [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+	_card_container.draw_cards(1)
+
+	var _queued_card: CardWorld = _card_container.cards_in_hand[0]
+	_queued_card.card_data.energy.energy_cost = 999
+
+	_card_container._on_card_clicked(_queued_card)
+
+	assert_false(_card_container.is_card_queued())
+
+func test_having_energy():
```
Comment (Turtyo) [inline Tests/test_energy.gd:40]: `test_play_card_with_enough_energy` ? to have energy doesn't mean you have enough

File: Tests/test_energy.gd [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+
+	_player_energy_component.add_energy(10)
```
Comment (Turtyo) [inline Tests/test_energy.gd:42]: I notice that you always use energy amounts that the player won't usually go to
it would be good to have some functions that test for energy levels that would happen more usually

File: Tests/test_energy.gd [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+func test_no_energy():
+
+	_card_container.draw_cards(1)
+
+	var _queued_card: CardWorld = _card_container.cards_in_hand[0]
+	_queued_card.card_data.energy.energy_cost = 999
+
+	_card_container._on_card_clicked(_queued_card)
+
+	assert_false(_card_container.is_card_queued())
```
Comment (Turtyo) [inline Tests/test_energy.gd:38]: could also assert that the energy didn't change

**Review by Turtyo — CHANGES_REQUESTED** [review]
I think the tests could cover more cases.
For example, testing when you have just enough energy for a card
testing if the energy you have is just one less than the energy of the card

File: Cards/CardBase.gd [hunk 18]
```diff
@@ -20,6 +20,8 @@ class_name CardBase
 
 @export var card_effects_data: Array[EffectData] = []
 
+@export var energy: EnergyData = EnergyData.new()
```
Comment (Turtyo) [inline Cards/CardBase.gd:23]: I would maybe name this something else than `energy`
because `energy` alone feels like this would be a cost (which is in fact `energy_cost`)
Maybe `energy_info` would be more fitting (so we know it's information and we need to query the actual things in it)

File: #Scenes/TestingScene.tscn
Comment (cheesefrycook) [inline #Scenes/TestingScene.tscn:1]: We should probably switch the energy UI display image to a TextureRect instead of a Sprite2D

File: Cards/CardBase.gd [hunk 19]
```diff
@@ -46,6 +48,9 @@ func on_card_play(caster: Entity, targets: Array[Entity]) -> void:
 	#Split up targeted attacks and all attacks
 	var effects_targeting_single: Array[EffectData] = get_effects_targeting_single()
 	var effects_targeting_all: Array[EffectData] = get_effects_target_all()
+
+	#uses the energy asocieted with the card
+	PlayerManager.player.get_energy_component().use_energy(self)
```
Comment (cheesefrycook) [inline Cards/CardBase.gd:53]: Maybe add a `if caster is Player` before this call, since enemies use `on_card_play` too

File: Entity/Components/EnergyComponent.gd [hunk 20]
```diff
@@ -0,0 +1,43 @@
...
+
+	energy += energy_generation
+
+	if energy >= Max_energy:
+		energy = Max_energy
+
+	on_energy_changed.emit()
+
+func add_energy(value: int) -> void:
+	energy += value
```
Comment (cheesefrycook) [inline Entity/Components/EnergyComponent.gd:42]: So if you add energy from a card then it goes above the `max_energy`? Is that by design?

File: Entity/Components/EnergyComponent.gd (hunk 10 above)
Comment (cheesefrycook) [reply Entity/Components/EnergyComponent.gd:32]: There's already a place to do this too. CardContainer -> _on_player_start_turn

File: Entity/Components/EnergyComponent.gd (hunk 11 above)
Comment (Paper-2) [reply Entity/Components/EnergyComponent.gd:43]: I forgot to delete the print.

File: Entity/Components/EnergyComponent.gd (hunk 20 above)
Comment (Paper-2) [reply Entity/Components/EnergyComponent.gd:42]: The energy only returns to the maximum when the turn ends (or starts. after I change it)

File: Entity/Components/EnergyComponent.gd (hunk 10 above)
Comment (Turtyo) [reply Entity/Components/EnergyComponent.gd:32]: yes but your function is `on_turn_end`
do you plan to call `on_turn_end` inside `_on_player_turn_start` ?

File: Cards/CardBase.gd (hunk 19 above)
Comment (Turtyo) [reply Cards/CardBase.gd:53]: good catch

File: Entity/Components/EnergyComponent.gd (hunk 20 above)
Comment (Turtyo) [reply Entity/Components/EnergyComponent.gd:42]: It's indeed by design. You can check the explanation on the related issue #56

File: Cards/Resource/Card_DrawCards.tres [hunk 21]
```diff
@@ -14,14 +13,9 @@ effect = SubResource("Resource_2ap55")
 value = 2
 target_type = 1
 
-[sub_resource type="Resource" id="Resource_j6isw"]
-script = ExtResource("3_06oqv")
-energy_cost = 1
```
Comment (Turtyo) [inline Cards/Resource/Card_DrawCards.tres]: does that mean the DrawCard doesn't cost anything ? wouldn't it be better to have the energy cost of 0 to not try to get the energy info and getting null instead ?

File: Cards/Resource/Card_Energy.tres [hunk 22]
```diff
@@ -14,14 +14,14 @@ effect = SubResource("Resource_jur0i")
 value = 70
 target_type = 0
 
-[sub_resource type="Resource" id="Resource_1100t"]
-script = ExtResource("3_2mfb5")
-energy_cost = 0
```
Comment (Turtyo) [inline Cards/Resource/Card_Energy.tres]: so the card that gives you energy costs 2 energy ?

File: Tests/test_energy.gd [hunk 23]
```diff
@@ -8,42 +8,67 @@ func before_each():
...
+	card.energy_info.energy_cost = 0
 	card.on_card_play(_player, [_player])
 	assert_eq(_player_energy_component.energy, 70)
 
-
 func test_using_energy():
 	_player_energy_component.add_energy(10)
-	card = load("res://Cards/Resource/Card_Damage.tres") # costs 7 by default
-	card.energy.energy_cost = 7
+	card = load("res://Cards/Resource/Card_Damage.tres")
```
Comment (Turtyo) [inline Tests/test_energy.gd:21]: missing type of the card (CardBase)

**Review by Turtyo — COMMENTED** [review]
Good changes, I have a few more questions

File: Tests/test_energy.gd (hunk 23 above)
Comment (Paper-2) [reply Tests/test_energy.gd:21]: Line 3

File: Cards/Resource/Card_DrawCards.tres (hunk 21 above)
Comment (Paper-2) [reply Cards/Resource/Card_DrawCards.tres]: if the value isn't set then it will use the one set in cardbase. (when I renamed energy to energy info I had to reset the values for every card I may or may not have skipped it)

Comment (Turtyo) [reply Cards/Resource/Card_DrawCards.tres]: I see, default is 0

File: Cards/Resource/Card_Energy.tres (hunk 22 above)
Comment (Paper-2) [reply Cards/Resource/Card_Energy.tres]: Can you rephrase this?

Comment (Turtyo) [reply Cards/Resource/Card_Energy.tres]: Nvm I saw the card, it gives you 70 energy for a cost of 2
I was just surprised that the card that adds to your energy costs energy, but since it gives a lot it doesn't matter
this will be changed later to a more reasonable value

File: Cards/Resource/Card_Energy.tres [hunk 24]
```diff
@@ -0,0 +1,27 @@
...
+[ext_resource type="Script" path="res://Cards/CardBase.gd" id="3_opvsg"]
+
+[sub_resource type="Resource" id="Resource_jur0i"]
+script = ExtResource("1_3yh17")
+
+[sub_resource type="Resource" id="Resource_6gvxm"]
+script = ExtResource("2_dr7f0")
+effect = SubResource("Resource_jur0i")
+value = 70
+target_type = 0
```
Comment (Turtyo) [inline Cards/Resource/Card_Energy.tres:15]: this is not insta-cast
things will have to change with the targeting PR anyway, i will rewrite this card once everything is merged I think

Comment (Paper-2) [reply Cards/Resource/Card_Energy.tres:15]: Just for reference, what is insta-cast?

Comment (Turtyo) [reply Cards/Resource/Card_Energy.tres:15]: Insta-cast is the mechanic that makes it so you don't need to target someone when it's clear who the card is for (buff, heal, damage everyone etc...)
This changed due to the targeting PR that was just merged, ask me if you don't understand the new system

File: UI/EnergyLabel.gd [hunk 25]
```diff
@@ -0,0 +1,20 @@
...
+
+var energy_component: EnergyComponent
+
+# Called when the node enters the scene tree for the first time.
+func _ready():
```
Comment (Turtyo) [inline UI/EnergyLabel.gd:6]: All the functions in this file are missing a return type (which causes errors when loading the game, since missing typing is treated as an error now)

File: UI/EnergyLabel.gd [hunk 26]
```diff
@@ -0,0 +1,20 @@
...
+
+# Called when the node enters the scene tree for the first time.
+func _ready() -> void:
+	energy_component = PlayerManager.player.get_energy_component()
+	energy_component.on_energy_changed.connect(on_energy_changed)
+	CardManager.on_card_action_finished.connect(on_card_played)
+	update_label()
+
+
+func on_card_played(_card: CardWorld) -> void:
```
Comment (Turtyo) [inline UI/EnergyLabel.gd:13]: I was also notice that you have the function with a card parameter that is not used
I understand it is because the `on_card_action_finished` returns card. Instead of having a parameter that you don't use, I suggest doing:
`CardManager.on_card_action_finished.connect(on_card_played.unbind(1))` which will basically ignore the value that is given by `on_card_action_finished`
Because right now I was wondering if `on_card_player` was an override or something like that

**Review by Turtyo — APPROVED** [review]

**Review by cheesefrycook — APPROVED** [review]
